eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Handler.php

```diff
@@ -628,46 +639,122 @@ public function locationSwapped($location1Id, $location1ParentId, $location2Id,
         $autoLocation1 = $this->gateway->loadAutogeneratedEntry("eznode:{$location1Id}");
         $autoLocation2 = $this->gateway->loadAutogeneratedEntry("eznode:{$location2Id}");
 
+        $contentInfo1 = $this->contentGateway->loadContentInfoByLocationId($location1Id);
+        $contentInfo2 = $this->contentGateway->loadContentInfoByLocationId($location2Id);
+
+        $names1 = $this->getNamesForAllLanguages($contentInfo1);
+        $names2 = $this->getNamesForAllLanguages($contentInfo2);
+
+        $alwaysAvailable1 = (bool)($contentInfo1['language_mask'] & 1);
+        $alwaysAvailable2 = (bool)($contentInfo2['language_mask'] & 1);
+
+        $languages = $this->languageHandler->loadAll();
+
         // Historize everything first to avoid name conflicts in case swapped Locations are siblings
         $this->historizeBeforeSwap($location1Entries, $location2Entries);
 
-        foreach ($location2Entries as $row) {
-            $alwaysAvailable = (bool)($row['lang_mask'] & 1);
-            $languageIds = $this->extractLanguageIdsFromMask($row['lang_mask']);
+        foreach ($languages as $languageCode => $language) {
+            $isMainLanguage1 = $language->id == $location1MainLanguageId;
+            $isMainLanguage2 = $language->id == $location2MainLanguageId;
+            if (isset($names1[$languageCode], $names2[$languageCode])) {
+                $name1 = $names1[$languageCode];
+                $name2 = $names2[$languageCode];
+                if ($this->shouldUrlAliasForSecondLocationBePublishedFirst(
```

Comment:
This whole loop contains too many code repetitions. What you need here is a method that returns an ordered list of sets of arguments to be passed to `internalPublishUrlAliasForLocation`. It could be some internal DTO ValueObject or just an associative array.

Maybe something like (adjust naming):
```php
$locationUrlAliasesList = $this->geUrlAliasesToBePublishedForLocations();
```
and internally instead of checking `shouldUrlAliasForSecondLocationBePublishedFirst` just "inline" that code and put proper data on the resulting list in a proper order. Don't forget about docblock describing what `shouldUrlAliasForSecondLocationBePublishedFirst` does, It's great :)

Please contact me if something here is not clear.